tensorflow/core/kernels/svd_op_gpu.cu.cc

```diff
@@ -0,0 +1,427 @@
+#if GOOGLE_CUDA
+#define EIGEN_USE_GPU
+
+#include <algorithm>
+#include <vector>
+
+#include "tensorflow/core/framework/kernel_def_builder.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/kernels/linalg_ops_common.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/platform/logging.h"
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/core/kernels/cuda_solvers.h"
+#include "tensorflow/core/platform/stream_executor.h"
+#include "tensorflow/core/util/cuda_kernel_helper.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
+// I need to transpose V afterwards
+#include "tensorflow/core/kernels/transpose_functor.h"
+
+// Logging
+#include <stdio.h>
+
+namespace tensorflow {
+
+static const char kErrMsg[] =
+    "Singular Value Decomposition was not successful. The input might not be "
+    "valid.";
+
+typedef Eigen::GpuDevice GPUDevice;
+
+namespace {
+  // This kernel computes the reduction
+  // V' = sum_i (M_i * U_i,1 * S_i).
+  // The result is stored in V[batch] and has the same sign as the
+  // real value of V (which should be computed)  
+  template<class Scalar>
+  __global__ void ComputeValueOfVKernel(CudaLaunchConfig config,
+      int64 m, const Scalar* M, const Scalar* U, const Scalar* S, Scalar* V)
+  {
+    CUDA_1D_KERNEL_LOOP(i, config.virtual_thread_count) {
+      Scalar v = M[i] * U[m*i] * S[0];
+      CudaAtomicAdd(V, v);
+    }
+  }
+
+  // Extracts the sign of V
+  // V[i] = V[i]>=0 ? 1 : 0
+  template<class Scalar>
+  __global__ void ExtractSignOfVKernel(CudaLaunchConfig config, Scalar* V)
+  {
+    CUDA_1D_KERNEL_LOOP(i, config.virtual_thread_count) {
+      V[i] = V[i]>=0 ? (Scalar)1 : (Scalar)0;
+    }
+  }
+}
+
+// Scalar: The input scalar type (can be complex)
+template <class Scalar>
+class SvdOpGpu : public AsyncOpKernel {
+ public:
+  using SScalar = typename Eigen::NumTraits<Scalar>::Real;
+ 
+  explicit SvdOpGpu(OpKernelConstruction* context) : AsyncOpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("compute_uv", &compute_uv_));
+    OP_REQUIRES_OK(context, context->GetAttr("full_matrices", &full_matrices_));
+  }
+
+  void RunSVD(OpKernelContext* context, DoneCallback done, int64 m, int64 n,
+              int64 p, int64 batch_size, Scalar* input_ptr,
+              SScalar* outputS_ptr, Scalar* outputU_ptr, Scalar* outputVT_ptr,
+              int* dev_info_ptr, CudaSolver& solver) {
+
+    // Needed for the n=1 fix, see below, since SVD destroys the input
+    Tensor input_copy;
+    if (compute_uv_ && n==1) {
+      OP_REQUIRES_OK_ASYNC(
+        context,
+        context->allocate_temp(DataTypeToEnum<Scalar>::v(), TensorShape({m}), &input_copy),
+        done);
+    }
+
+    for (int64 batch = 0; batch < batch_size; ++batch) {
+      int lda = m;
```

Comment:
You don't really need these temporary variables for leading dimensions, since they are passed to solver.Gesvd(...) by value. (This is not Fortran ;-) )